File: tests/fixtures/taint_callback.js
// Taint flow patterns: callbacks, loops, sanitizers

function forEachTaint(req, res) {
    const items = req.body.items;
    items.forEach(item => {
        res.send("<div>" + item + "</div>");
    });
}

function mapTaint(req, res) {
    const names = req.query.names;
    const html = names.map(name => "<li>" + name + "</li>");
    res.send(html.join(""));
}

function mapSanitized(req, res) {
    const names = req.query.names;
    const html = names.map(name => "<li>" + escapeHtml(name) + "</li>");
    res.send(html.join(""));
}

function forOfTaint(req, res) {
    const entries = req.body.entries;
    for (const entry of entries) {
        eval(entry);
    }
}

function forOfDestructure(req, res) {
    const entries = req.body.entries;
    for (const { key, value } of entries) {
        res.send(key + "=" + value);
    }
}

function templateSink(req, res) {
    const name = req.query.name;
    document.getElementById("out").innerHTML = `<h1>${name}</h1>`;
}

function noTaint() {
    const items = ["safe", "values"];
    items.forEach(item => {
        console.log(item);
    });
}

function filterTaint(req, res) {
    const users = req.body.users;
    const active = users.filter(u => u.active);
    res.json(active);
}

File: tests/fixtures/taint_callback.py
# Taint flow patterns: for-loops, map, comprehensions

def for_loop_taint(request):
    items = request.args.getlist("items")
    for item in items:
        cursor.execute(f"SELECT * FROM t WHERE name = '{item}'")

def map_taint(request):
    names = request.args.getlist("names")
    queries = list(map(lambda n: f"SELECT * FROM users WHERE name = '{n}'", names))
    for q in queries:
        cursor.execute(q)

def comprehension_taint(request):
    ids = request.args.getlist("ids")
    queries = [f"DELETE FROM t WHERE id = {i}" for i in ids]
    for q in queries:
        cursor.execute(q)

def for_loop_sanitized(request):
    items = request.args.getlist("items")
    for item in items:
        safe = escape(item)
        output(f"<div>{safe}</div>")

def nested_function_taint(request):
    data = request.form.get("data")
    def process():
        return eval(data)
    process()
